File: packages/protocol/src/semantic-root.ts
import { encodeCanonical, FormatError, type CborValue } from "./cbor.js";
import { formatArtifactId, parseArtifactId } from "./artifact.js";
import { validatePath } from "./path.js";
import type { Realm } from "./realm.js";

export interface SemanticArtifact {
  id: string;
  realm: Realm;
}

export interface SemanticRef {
  name: string;
  target: string;
  realm: Realm;
}

export interface SemanticRootInput {
  project: string;
  realm: Realm;
  artifacts: readonly SemanticArtifact[];
  refs: readonly SemanticRef[];
  policyVersion: bigint;
}

export interface SemanticRootResult {
  artifactSetRoot: Uint8Array;
  descriptor: Uint8Array;
  semanticRoot: string;
}

const MAX_REALM_ARTIFACTS = 65_535;
const MAX_REALM_REFS = 4_096;
const textEncoder = new TextEncoder();

function schemaError(message: string): never {
  throw new FormatError("invalid_schema", message);
}

function compareBytes(left: Uint8Array, right: Uint8Array): number {
  const shared = Math.min(left.length, right.length);
  for (let index = 0; index < shared; index += 1) {
    const difference = left[index]! - right[index]!;
    if (difference !== 0) return difference;
  }
  return left.length - right.length;
}

async function sha256(bytes: Uint8Array): Promise<Uint8Array> {
  const buffer = new ArrayBuffer(bytes.byteLength);
  new Uint8Array(buffer).set(bytes);
  return new Uint8Array(await crypto.subtle.digest("SHA-256", buffer));
}

function validateRefName(name: string): void {
  if (textEncoder.encode(name).length > 255) {
    schemaError("ref name exceeds 255 UTF-8 bytes");
  }
  try {
    validatePath(name);
  } catch {
    schemaError("ref name violates edgefossil-path-v0");
  }
}

/** Computes one realm's portable semantic root while ignoring other realms. */
export async function computeSemanticRoot(
  input: SemanticRootInput,
): Promise<SemanticRootResult> {
  const project = parseArtifactId(input.project);
  if (
    input.policyVersion < 0n ||
    input.policyVersion > 0xffff_ffff_ffff_ffffn
  ) {
    schemaError("policy_version must be uint64");
  }

  const selectedArtifacts = input.artifacts
    .filter(({ realm }) => realm === input.realm)
    .map(({ id }) => ({ id, digest: parseArtifactId(id) }));
  if (selectedArtifacts.length > MAX_REALM_ARTIFACTS) {
    throw new FormatError("resource_limit", "realm artifact set is too large");
  }
  selectedArtifacts.sort((left, right) =>
    compareBytes(left.digest, right.digest),
  );
  for (let index = 1; index < selectedArtifacts.length; index += 1) {
    if (
      compareBytes(
        selectedArtifacts[index - 1]!.digest,
        selectedArtifacts[index]!.digest,
      ) === 0
    ) {
      schemaError("realm artifact set contains a duplicate");
    }
  }
  const selectedIds = new Set(selectedArtifacts.map(({ id }) => id));
  const artifactSetBytes = encodeCanonical(
    selectedArtifacts.map(({ digest }) => digest),
  );
  const artifactSetRoot = await sha256(artifactSetBytes);

  const selectedRefs = input.refs.filter(({ realm }) => realm === input.realm);
  if (selectedRefs.length > MAX_REALM_REFS) {
    throw new FormatError("resource_limit", "realm ref set is too large");
  }
  const refs = new Map<string, CborValue>();
  for (const ref of selectedRefs) {
    validateRefName(ref.name);
    if (refs.has(ref.name)) schemaError("realm ref name is duplicated");
    const target = parseArtifactId(ref.target);
    if (!selectedIds.has(ref.target)) {
      throw new FormatError(
        "unknown_required_semantics",
        "realm ref target is not in the realm artifact set",
      );
    }
    refs.set(ref.name, target);
  }

  const descriptor = encodeCanonical(
    new Map<string, CborValue>([
      ["format", "edgefossil-semantic-root"],
      ["version", 0n],
      ["project", project],
      ["realm", input.realm],
      ["artifact_set_root", artifactSetRoot],
      ["refs", refs],
      ["policy_version", input.policyVersion],
    ]),
  );
  return {
    artifactSetRoot,
    descriptor,
    semanticRoot: formatArtifactId(await sha256(descriptor)),
  };
}

File: crates/ef-format/src/semantic_root.rs
//! Realm-isolated portable semantic-root calculation.

use std::collections::HashSet;

use sha2::{Digest, Sha256};

use crate::{
    FormatError, FormatErrorCode, Realm, Value, encode_value, format_artifact_id,
    parse_artifact_id, validate_path,
};

const MAX_REALM_ARTIFACTS: usize = 65_535;
const MAX_REALM_REFS: usize = 4_096;

/// One immutable artifact candidate for a semantic-root view.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SemanticArtifact {
    pub id: String,
    pub realm: Realm,
}

/// One named ref candidate for a semantic-root view.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SemanticRef {
    pub name: String,
    pub target: String,
    pub realm: Realm,
}

/// Inputs for one realm's semantic root.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SemanticRootInput {
    pub project: String,
    pub realm: Realm,
    pub artifacts: Vec<SemanticArtifact>,
    pub refs: Vec<SemanticRef>,
    pub policy_version: u64,
}

/// Reproducible intermediate and final semantic-root values.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SemanticRootResult {
    pub artifact_set_root: [u8; 32],
    pub descriptor: Vec<u8>,
    pub semantic_root: String,
}

fn schema_error(message: &str) -> FormatError {
    FormatError::new(FormatErrorCode::InvalidSchema, message)
}

fn validate_ref_name(name: &str) -> Result<(), FormatError> {
    if name.len() > 255 {
        return Err(schema_error("ref name exceeds 255 UTF-8 bytes"));
    }
    validate_path(name).map_err(|_| schema_error("ref name violates edgefossil-path-v0"))
}

/// Computes one realm's portable semantic root while ignoring other realms.
///
/// # Errors
///
/// Returns a stable format error for invalid IDs, duplicate selected inputs,
/// invalid ref names, missing selected ref targets, or resource limits.
pub fn compute_semantic_root(input: &SemanticRootInput) -> Result<SemanticRootResult, FormatError> {
    let project = parse_artifact_id(&input.project)?;
    let mut selected_artifacts = input
        .artifacts
        .iter()
        .filter(|artifact| artifact.realm == input.realm)
        .map(|artifact| Ok((artifact.id.as_str(), parse_artifact_id(&artifact.id)?)))
        .collect::<Result<Vec<_>, FormatError>>()?;
    if selected_artifacts.len() > MAX_REALM_ARTIFACTS {
        return Err(FormatError::new(
            FormatErrorCode::ResourceLimit,
            "realm artifact set is too large",
        ));
    }
    selected_artifacts.sort_by(|left, right| left.1.cmp(&right.1));
    if selected_artifacts
        .windows(2)
        .any(|pair| pair[0].1 == pair[1].1)
    {
        return Err(schema_error("realm artifact set contains a duplicate"));
    }
    let selected_ids = selected_artifacts
        .iter()
        .map(|(id, _)| *id)
        .collect::<HashSet<_>>();
    let artifact_set_bytes = encode_value(&Value::Array(
        selected_artifacts
            .iter()
            .map(|(_, digest)| Value::Bytes(digest.to_vec()))
            .collect(),
    ))?;
    let artifact_set_root: [u8; 32] = Sha256::digest(artifact_set_bytes).into();

    let selected_refs = input
        .refs
        .iter()
        .filter(|reference| reference.realm == input.realm)
        .collect::<Vec<_>>();
    if selected_refs.len() > MAX_REALM_REFS {
        return Err(FormatError::new(
            FormatErrorCode::ResourceLimit,
            "realm ref set is too large",
        ));
    }
    let mut seen_ref_names = HashSet::new();
    let mut refs = Vec::with_capacity(selected_refs.len());
    for reference in selected_refs {
        validate_ref_name(&reference.name)?;
        if !seen_ref_names.insert(reference.name.as_str()) {
            return Err(schema_error("realm ref name is duplicated"));
        }
        let target = parse_artifact_id(&reference.target)?;
        if !selected_ids.contains(reference.target.as_str()) {
            return Err(FormatError::new(
                FormatErrorCode::UnknownRequiredSemantics,
                "realm ref target is not in the realm artifact set",
            ));
        }
        refs.push((reference.name.clone(), Value::Bytes(target.to_vec())));
    }

    let descriptor = encode_value(&Value::Map(vec![
        (
            "format".into(),
            Value::Text("edgefossil-semantic-root".into()),
        ),
        ("version".into(), Value::UInt(0)),
        ("project".into(), Value::Bytes(project.to_vec())),
        ("realm".into(), Value::Text(input.realm.as_str().into())),
        (
            "artifact_set_root".into(),
            Value::Bytes(artifact_set_root.to_vec()),
        ),
        ("refs".into(), Value::Map(refs)),
        ("policy_version".into(), Value::UInt(input.policy_version)),
    ]))?;
    let semantic_digest: [u8; 32] = Sha256::digest(&descriptor).into();
    let semantic_root = format_artifact_id(&semantic_digest);
    Ok(SemanticRootResult {
        artifact_set_root,
        descriptor,
        semantic_root,
    })
}
